homeassistant/helpers/trigger.py

```diff
@@ -49,12 +50,37 @@
 )
 
 
+class Trigger(abc.ABC):
+    """Trigger class."""
+
+    @classmethod
+    @abc.abstractmethod
+    async def async_validate_trigger_config(
+        cls, hass: HomeAssistant, config: ConfigType
+    ) -> ConfigType:
+        """Validate config."""
+
+    @classmethod
+    @abc.abstractmethod
+    async def async_attach_trigger(
+        cls,
+        hass: HomeAssistant,
+        config: ConfigType,
+        action: TriggerActionType,
+        trigger_info: TriggerInfo,
+    ) -> CALLBACK_TYPE:
+        """Attach a trigger."""
```

Comment:
I think making the validator a class method makes sense.
However, I think the `async_attach_trigger` should either be a normal method, or it should create an instance.
In that way, `async_attach_trigger` can store data in the trigger object itself if needed.

Reply:
Yeah, I think we should allow or probably require instantiating the class.

Reply:
Instantiating can be done from the `async_attach_trigger` class method if needed, this PR shows how it can work: https://github.com/home-assistant/core/pull/144885

Reply:
I and @MartinHjelmare did explore making `async_attach_trigger` an instance method, but considering these points we concluded a classmethod is more flexible:
- There's nothing which will instantiate a trigger and keep a reference to it without also attaching it
- We don't want the `__init__` method to have the side effect of attaching the trigger

Reply:
I don't think it's a sane interface if a method can either be a class method or an instance method, so if we change we should make it an instance method always.

The rationale for the class method is that nothing needs to keep a reference to trigger instances.
An instance method would allow keeping instances to the actual trigger instances, which is much more flexible and leaves the door open for future extensions of the trigger class. OTOH, YAGNI 🤷 

How is the class method not nice from a developer experience?